File: javascript/json-transform/src/formats/csv/CsvFormat.ts
import { getAsString, isMap, isNullOrUndefined } from "../../JsonHelpers";
import { FormatSerializer } from "../FormatSerializer";
import { FormatDeserializer } from "../FormatDeserializer";

const MIN_SUPPLEMENTARY_CODE_POINT = 0x010000;
function charCount(codePoint: number) {
  return codePoint >= MIN_SUPPLEMENTARY_CODE_POINT ? 2 : 1;
}

class CsvFormat implements FormatSerializer, FormatDeserializer {
  private static readonly COMMA = ",";
  private static readonly DEFAULT_SEPARATOR = CsvFormat.COMMA;
  private static readonly DOUBLE_QUOTES = '"';
  private static readonly EMBEDDED_DOUBLE_QUOTES = '""';
  private static readonly NEW_LINE_UNIX = "\n";
  private static readonly LINE_FEED = "\n".codePointAt(0);
  private static readonly CARRIAGE_RETURN = "\r".codePointAt(0);
  private static readonly NEW_LINE_WINDOWS = "\r\n";

  /**
   * Formatter -
   * if names provided it will use it as header row (unless noHeaders) and expose only those fields (if objects)
   * if names not provided, it will use the names of the first object and expose all those fields for the rest
   * Parser -
   * if names provided it will extract only those fields
   * if noHeaders is set, it will use these names to map the row values by their indices
   */
  private readonly names?: string[];
  /**
   * When used in formatter, it means the headers row will not be written
   * When used in parser, it means that the output of parsing will be an array of arrays (unless names is specified)
   */
  private readonly noHeaders: boolean;
  /**
   * Formatting will quote all values
   */
  private readonly forceQuote: boolean;
  /**
   * What separator to format with or expect to parse with
   */
  private readonly separator: string;

  constructor(
    names?: string[] | null,
    noHeaders?: boolean | null,
    forceQuote?: boolean | null,
    separator?: string | null,
  ) {
    this.names = names ?? undefined;
    this.noHeaders = isNullOrUndefined(noHeaders) ? false : noHeaders;
    this.forceQuote = isNullOrUndefined(forceQuote) ? false : forceQuote;
    this.separator = isNullOrUndefined(separator) ? CsvFormat.DEFAULT_SEPARATOR : separator;
  }

  private appendEscaped(sb: StringBuilder, val: any): void {
    let value: string;
    if (val === null || val === undefined) {
      value = "";
    } else {
      value = getAsString(val) ?? "";
    }
    if (
      this.forceQuote ||
      value.includes(CsvFormat.COMMA) ||
      value.includes(CsvFormat.DOUBLE_QUOTES) ||
      value.includes(CsvFormat.NEW_LINE_UNIX) ||
      value.includes(CsvFormat.NEW_LINE_WINDOWS) ||
      value.startsWith(" ") ||
      value.endsWith(" ")
    ) {
      sb.append(CsvFormat.DOUBLE_QUOTES);
      sb.append(value.replace(new RegExp(CsvFormat.DOUBLE_QUOTES, "g"), CsvFormat.EMBEDDED_DOUBLE_QUOTES));
      sb.append(CsvFormat.DOUBLE_QUOTES);
    } else {
      sb.append(value);
    }
  }

  private appendHeaders(sb: StringBuilder, headers: string[]): void {
    if (this.noHeaders) return;
    let first = true;
    for (const name of headers) {
      if (!first) {
        sb.append(this.separator);
      } else {
        first = false;
      }
      this.appendEscaped(sb, name);
    }
    sb.append("\n");
  }

  private appendRow(sb: StringBuilder, names: string[] | null | undefined, value: any): void {
    if (!Array.isArray(value) && names) {
      if (typeof value !== "object" || value === null) return;
      let first = true;
      for (const name of names) {
        if (!first) {
          sb.append(this.separator);
        } else {
          first = false;
        }
        this.appendEscaped(sb, value[name]);
      }
    } else {
      let first = true;
      for (const val of value) {
        if (!first) {
          sb.append(this.separator);
        } else {
          first = false;
        }
        this.appendEscaped(sb, val);
      }
    }
    sb.append("\n");
  }

  serialize(payload: any): string | null {
    const sb = new StringBuilder();
    let headers = this.names;
    if (headers) {
      this.appendHeaders(sb, headers);
    }

    if (Array.isArray(payload)) {
      if (!headers && payload.length > 0 && isMap(payload[0])) {
        headers = Object.keys(payload[0]);
        this.appendHeaders(sb, headers);
      }
      for (const x of payload) {
        this.appendRow(sb, headers, x);
      }
    } else {
      throw new Error("Unsupported object type to be formatted as CSV");
    }

    return sb.toString();
  }

  private accumulate(context: CsvParserContext, results: any[], values: any[]): void {
    if (results.length === 0 && !context.namesRead && !this.noHeaders) {
      context.names = values;
      context.namesRead = true;
      return;
    }
    if (this.noHeaders && isNullOrUndefined(this.names)) {
      results.push(values); // set item as an array of values
      return;
    }
    // there are names, make a map
    if (!isNullOrUndefined(context.names)) {
      const item: Record<string, any> = {};
      let i: number;
      for (i = 0; i < context.names.length; i++) {
        const name = getAsString(context.names[i]) ?? "";
        if ((isNullOrUndefined(context.extractNames) || context.extractNames.includes(name)) && values.length > i) {
          item[name] = values[i];
        }
      }
      // TODO: make it conditional
      // more values than names
      for (; i < values.length; i++) {
        if (!Object.prototype.hasOwnProperty.call(item, `$${i}`)) {
          item[`$${i}`] = values[i];
        }
      }
      results.push(item);
    }
  }

  deserialize(input: string | null): any {
    if (input === null) {
      return null;
    }
    const result: any[] = [];
    const context = new CsvParserContext();
    if (this.noHeaders && !isNullOrUndefined(this.names)) {
      context.names = this.names.slice();
    }
    context.extractNames = this.names ?? null;

    const len = input.length;
    let row: any[] = [];
    const cell = new StringBuilder();
    let offset = 0;

    while (offset < len) {
      // always take current one and next one (offset may skip 1 or 2 depending on char sequence)
      const cur = input.codePointAt(offset) as number;
      const curSize = charCount(cur);
      const next = offset + curSize < len ? (input.codePointAt(offset + curSize) as number) : -1;
      const curAndNextSize = curSize + charCount(next);

      if (cur === this.separator.codePointAt(0)) {
        if (context.inQuotes) {
          cell.append(this.separator);
        } else {
          row.push(cell.toString());
          cell.clear();
        }

        offset += curSize;
      } else if ((cur === CsvFormat.CARRIAGE_RETURN && next === CsvFormat.LINE_FEED) || cur === CsvFormat.LINE_FEED) {
        const unix = cur === CsvFormat.LINE_FEED;
        const eof = offset + (unix ? curSize : curAndNextSize) === len;
        if (!eof) {
          if (context.inQuotes) {
            cell.append(unix ? CsvFormat.NEW_LINE_UNIX : CsvFormat.NEW_LINE_WINDOWS);
          } else {
            row.push(cell.toString());
            cell.clear();
            this.accumulate(context, result, row);
            row = [];
          }
        }
        offset += unix ? curSize : curAndNextSize;
      } else if (cur === 34 && next === 34) {
        if (context.inQuotes) {
          cell.append(CsvFormat.DOUBLE_QUOTES);
          offset += curAndNextSize;
        } else if (cell.length === 0) {
          // consider only the first one
          context.inQuotes = !context.inQuotes;
          offset += curSize;
        } else {
          cell.append(CsvFormat.DOUBLE_QUOTES);
          offset += curSize;
        }
      } else if (cur === 34) {
        context.inQuotes = !context.inQuotes;
        offset += curSize;
      } else if (!context.inQuotes && (cur === 32 || cur === 9)) {
        // ignore
        offset += curSize;
      } else {
        cell.append(String.fromCodePoint(cur));
        offset += curSize;
      }
    }

    if (result.length || cell.length > 0) {
      row.push(cell.toString());
      this.accumulate(context, result, row);
    }
    return result as any;
  }
}

class CsvParserContext {
  public inQuotes = false;
  public names: string[] | null = null;
  public namesRead = false;
  public extractNames: string[] | null = null;
}

class StringBuilder {
  private strings: string[] = [];

  public append(str: string): void {
    this.strings.push(str);
  }

  public toString(): string {
    return this.strings.join("");
  }

  public clear(): void {
    this.strings.length = 0;
  }

  public get length(): number {
    return this.toString().length;
  }
}

export default CsvFormat;

File: java/json-transform/src/main/java/co/nlighten/jsontransform/formats/csv/CsvFormat.java
package co.nlighten.jsontransform.formats.csv;

import co.nlighten.jsontransform.adapters.JsonAdapter;
import co.nlighten.jsontransform.formats.FormatDeserializer;
import co.nlighten.jsontransform.formats.FormatSerializer;

import java.lang.reflect.Array;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;

public class CsvFormat implements FormatSerializer, FormatDeserializer {
    private static final String COMMA = ",";
    private static final String DEFAULT_SEPARATOR = COMMA;
    private static final String DOUBLE_QUOTES = "\"";
    private static final String EMBEDDED_DOUBLE_QUOTES = "\"\"";
    private static final String NEW_LINE_UNIX = "\n";
    private static final char LINE_FEED = '\n';
    private static final char CARRIAGE_RETURN = '\r';
    private static final String NEW_LINE_WINDOWS = "\r\n";

    /**
     * Formatter -
     * if names provided it will use it as header row (unless noHeaders) and expose only those fields (if objects)
     * if names not provided, it will use the names of the first object and expose all those fields for the rest
     * Parser -
     * if names provided it will extract only those fields
     * if noHeaders is set, it will use these names to map the row values by their indices
     */
    final Collection<String> names;
    /**
     * When used in formatter, it means the headers row will not be written
     * When used in parser, it means that the output of parsing will be an array of arrays (unless names is specified)
     */
    final boolean noHeaders;
    /**
     * Formatting will quote all values
     */
    final boolean forceQuote;
    /**
     * What separator to format with or expect to parse with
     */
    final String separator;

    private final JsonAdapter<?, ?, ?> adapter;

    public CsvFormat(JsonAdapter<?, ?, ?> adapter, List<String> names, Boolean noHeaders, Boolean forceQuote, String separator) {
        this.adapter = adapter;
        this.names = names;
        this.noHeaders = noHeaders != null && noHeaders;
        this.forceQuote = forceQuote != null && forceQuote;
        this.separator = separator != null ? separator : DEFAULT_SEPARATOR;
    }

    public CsvFormat(JsonAdapter<?, ?, ?> adapter) {
        this(adapter, null, null, null, null);
    }

    private Iterable<?> asIterable(Object value) {
        if (adapter.isJsonObject(value)) {
            return null;
        }
        if (adapter.isJsonArray(value)) {
            return adapter.asIterable(value);
        }
        if (value instanceof Iterable<?> iter) {
            return iter;
        }
        if (value instanceof Object[] oa) {
            return Arrays.asList(oa);
        }
        return null;
    }

    public void appendEscaped(StringBuilder sb, Object val) {
        String value;
        if (val == null) {
            value = "";
        } else if (adapter.is(val)) {
            value = adapter.getAsString(val);
        } else {
            value = val.toString();
        }
        if (forceQuote ||
                value.contains(COMMA) ||
                value.contains(DOUBLE_QUOTES) ||
                value.contains(NEW_LINE_UNIX) ||
                value.contains(NEW_LINE_WINDOWS) ||
                value.startsWith(" ") ||
                value.endsWith(" ")) {
            sb.append(DOUBLE_QUOTES);
            sb.append(value.replace(DOUBLE_QUOTES, EMBEDDED_DOUBLE_QUOTES));
            sb.append(DOUBLE_QUOTES);
        } else {
            sb.append(value);
        }
    }

    public void appendHeaders(StringBuilder sb, Collection<String> headers) {
        if (noHeaders) return;
        var first = true;
        for (String name : headers) {
            if (!first) {
                sb.append(separator);
            } else {
                first = false;
            }
            appendEscaped(sb, name);
        }
        sb.append("\n");
    }

    public void appendRow(StringBuilder sb, Collection<String> names, Object value) {
        // check for special case of array of arrays
        Iterable<?> iter = asIterable(value);
        if (iter == null) {
            // this is a normal case of array of objects
            if (!adapter.isJsonObject(value))
                return;
            var first = true;
            for (String name : names) {
                if (!first) {
                    sb.append(separator);
                } else {
                    first = false;
                }
                appendEscaped(sb, adapter.get(value, name));
            }
        } else {
            var first = true;
            for (Object val : iter) {
                if (!first) {
                    sb.append(separator);
                } else {
                    first = false;
                }
                appendEscaped(sb, val);
            }
        }
        sb.append("\n");
    }

    @Override
    public String serialize(Object payload) {
        StringBuilder sb = new StringBuilder();

        var headers = names;
        if (headers != null) {
            appendHeaders(sb, headers);
        }

        // go through the items
        if (payload instanceof Iterable<?> iter) {
            if (headers == null) {
                var it = iter.iterator();
                if (it.hasNext()) {
                    var first = it.next();
                    if (first != null && adapter.isJsonObject(first)) {
                        headers = adapter.keySet(first);
                        appendHeaders(sb, headers);
                    }
                }
            }
            for (Object x : iter) {
                appendRow(sb, headers, x);
            }
        } else if (payload.getClass().isArray()) {
            var len = Array.getLength(payload);
            if (headers == null && len > 0) {
                var first = Array.get(payload, 0);
                if (first != null && adapter.isJsonObject(first)) {
                    headers = adapter.keySet(first);
                    appendHeaders(sb, headers);
                }
            }
            for (var i = 0; i < len; i++) {
                appendRow(sb, headers, Array.get(payload, i));
            }
        } else {
            throw new Error("Unsupported object type to be formatted as CSV");
        }

        return sb.toString();
    }

    private class CsvParserContext {
        public boolean inQuotes = false;
        public Object names = null;
        public boolean namesRead = false;
        public Collection<String> extractNames = null;
    }

    void accumulate(CsvParserContext context, Object results, Object values) {
        if (adapter.isEmpty(results) && !context.namesRead && !noHeaders) {
            context.names = values;
            context.namesRead = true;
            return;
        }
        if (noHeaders && names == null) {
            adapter.add(results, values); // set item as an array of values
            return;
        }
        // there are names, make a map
        if (context.names != null) {
            var item = adapter.createObject();
            int i;
            for (i = 0; i < adapter.size(context.names); i++) {
                var name = adapter.getAsString(adapter.get(context.names, i));
                if ((context.extractNames == null || context.extractNames.contains(name)) &&
                        adapter.size(values) > i) {
                    adapter.add(item, name, adapter.get(values, i));
                }
            }
            // TODO: make it conditional
            // more values than names
            for (; i < adapter.size(values); i++) {
                if (!adapter.has(item, "$" + i)) {
                    adapter.add(item, "$" + i, adapter.get(values, i));
                }
            }
            adapter.add(results, item);
        }
    }
    @Override
    public Object deserialize(String input) {
        var result = adapter.createArray();
        var context = new CsvParserContext();
        if (noHeaders && names != null) {
            context.names = adapter.createArray();
            names.forEach(item -> adapter.add(context.names, item));
        }
        context.extractNames = names;

        var len = input.length();
        var row = adapter.createArray();
        var cell = new StringBuilder();
        var offset = 0;

        while (offset < len) {
            // always take current one and next one (offset may skip 1 or 2 depending on char sequence)
            final int cur = input.codePointAt(offset);
            var curSize = Character.charCount(cur);
            final int next = offset + curSize < len ?  input.codePointAt(offset + curSize) : -1;
            var curAndNextSize = curSize + Character.charCount(next);

            if (cur == separator.codePointAt(0)) {
                if (context.inQuotes) {
                    cell.append(separator);
                } else {
                    adapter.add(row, cell.toString());
                    cell.setLength(0);
                }

                offset += curSize;
            } else if ( (cur == CARRIAGE_RETURN && next == LINE_FEED) || cur == LINE_FEED) {
                var unix = cur == LINE_FEED;
                var eof = offset + (unix ? curSize : curAndNextSize) == len;
                if (!eof) {
                    if (context.inQuotes) {
                        cell.append(unix ? NEW_LINE_UNIX : NEW_LINE_WINDOWS);
                    } else {
                        adapter.add(row, cell.toString());
                        cell.setLength(0);
                        accumulate(context, result, row);
                        row = adapter.createArray();
                    }
                }
                offset += unix ? curSize : curAndNextSize;
            } else if (cur == '"' && next == '"') { // span.startsWith(EMBEDDED_DOUBLE_QUOTES)) {
                if (context.inQuotes) {
                    cell.append(DOUBLE_QUOTES);
                    offset += curAndNextSize;
                } else if (cell.isEmpty()) {
                    // consider only the first one
                    context.inQuotes = !context.inQuotes;
                    offset += curSize;
                } else {
                    cell.append(DOUBLE_QUOTES);
                    offset += curSize;
                }
            } else if (cur == '"') { // span.startsWith(DOUBLE_QUOTES)) {
                context.inQuotes = !context.inQuotes;
                offset += curSize;
            } else if (!context.inQuotes && (cur == ' ' || cur == '\t')) { // (span.codePointAt(0) == ' ' || span.codePointAt(0) == '\t')) {
                // ignore
                offset += curSize;
            } else {
                cell.append(Character.toString(cur));
                offset += curSize;
            }
        }

        if (!adapter.isEmpty(result) || !cell.isEmpty()) {
            adapter.add(row, cell.toString());
            accumulate(context, result, row);
        }
        return result;
    }
}
